app/controllers/doorkeeper/discovery_controller.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module Doorkeeper
+  class DiscoveryController < Doorkeeper::ApplicationMetalController
+    def show
+      headers.merge!(discovery_response.headers)
+      render json: discovery_response.body,
+             status: discovery_response.status
+    end
+
+    private
+
+    def discovery_response
+      @discovery_response ||= Doorkeeper::OAuth::DiscoveryResponse.new(
+        root_url,
+        -> (**args) { url_for(**args) }
```

Comment:
I'm not particularly keen on this, but including UrlFor or similar in the `Doorkeeper::OAuth::DiscoveryResponse` class didn't work as I'd expect it to, and I need the url builder in order to generate the absolute URLs for the authorization server metadata.